R/score_untrained.R

```diff
@@ -0,0 +1,101 @@
+#' @title Calculate model importance for a single task when using an untrained
+#' ensemble
+#' @description
+#' Evaluate the importance of ensemble component models by quantifying their
+#' contribution to the prediction accuracy of an untrained ensemble for each
+#' combination of model task.
+#'
+#' @inheritParams model_importance
+#' @param single_task_data A data.frame with the predictions for a single
+#' forecast task in a model_out_tbl format. The data must contain only one
+#' `output_type`, which must be one of the following:
+#' `mean`, `median`, `quantile`, or `pmf`.
+#' @param model_id_list A list of all component model IDs to be used in the
+#' ensemble. If a model is not present in the `single_task_data`, it means that
+#' the model did not submit predictions for the given task.
+#' This list is used to identify missing models in the ensemble.
+#' @param metric A character string specifying the metric to be used for scoring
+#' the model output. The metric is determined by the `output_type` and must be
+#' one of the following: `se_point`, `ae_point`, `wis`, or `log_score`.
+#'
+#' @returns A data.frame with columns
+#' `task_id`, `output_type`, `model_id`, `task_level_importance`.
+#'
+#' @import hubEnsembles
+#' @import hubEvals
+#' @inherit model_importance details
+
+score_untrained <- function(single_task_data, oracle_output_data, model_id_list,
+                            ensemble_fun, importance_algorithm, subset_wt,
+                            metric, ...) {
+  # models in the single_task_data
+  models <- unique(single_task_data$model_id)
+  missing_model <- setdiff(model_id_list, single_task_data$model_id)
+  # Compute importance score when importance_algorithm is 'lomo'
+  if (importance_algorithm == "lomo") {
+    ens_all <- switch(ensemble_fun,
+      "simple_ensemble" = hubEnsembles::simple_ensemble(single_task_data,
+        weights = NULL,
+        model_id = "ensemble-all"
+      ),
+      "linear_pool" = hubEnsembles::linear_pool(single_task_data,
+        weights = NULL,
+        model_id = "ensemble-all"
+      )
+    )
+
+    # build ensemble forecasts by leaving one model out
+    ens_lomo <- lapply(models, function(x) {
+      single_task_data_lomo <- single_task_data |>
+        dplyr::filter(.data$model_id != x)
+      switch(ensemble_fun,
+        "simple_ensemble" = hubEnsembles::simple_ensemble(single_task_data_lomo,
+          weights = NULL,
+          model_id = paste0("ens.wo.", x)
+        ),
+        "linear_pool" = hubEnsembles::linear_pool(single_task_data_lomo,
+          weights = NULL,
+          model_id = paste0("ens.wo.", x)
+        )
+      )
```

Comment:
A small note about the switch statements you have here: since the arguments of both function calls are the same, you can extract the function from the `hubEnsembles` package and apply it to reduce duplication. 

```suggestion
    # `ENS_FUN` is a function from hubEnsembles specified by `ensemble_fun` 
    ENS_FUN <- getFromNamespace(ensemble_fun, ns = asNamespace("hubEnsembles"))
    ens_all <- ENS_FUN(single_task_data,
      weights = NULL,
      model_id = "ensemble-all"
    )

    # build ensemble forecasts by leaving one model out
    ens_lomo <- lapply(models, function(x) {
      single_task_data_lomo <- single_task_data |>
        dplyr::filter(.data$model_id != x) |>
        ENS_FUN(
          weights = NULL,
          model_id = paste0("ens.wo.", x)
        )
```

Reply:
@zkamvar Wow, this is fantastic! Thank you for it!